### PR TITLE
[SPARK-48873][SQL] Use UnsafeRow in JSON parser.

### PR DESCRIPTION
**Pull request by chenhao-db**
### What changes were proposed in this pull request?

It uses `UnsafeRow` to represent struct result in the JSON parser. It saves memory compared to the current `GenericInternalRow`. The change is guarded by a flag and disabled by default.

The benchmark shows that enabling the flag brings ~10% slowdown. This is basically expected because converting to `UnsafeRow` requires some work. The purpose of the PR is to provide an alternative to save memory.

I did the following experiment. It generates a big `.gz` JSON file containing a single large array. Each array element is a struct with 50 string fields and will be parsed into a row by the JSON reader.

```
s = b'{"field00":null,"field01":"field01_<v>","field02":"field02_<v>","field03":"field03_<v>","field04":"field04_<v>","field05":"field05_<v>","field06":"field06_<v>","field07":"field07_<v>","field08":"field08_<v>","field09":"field09_<v>","field10":null,"field11":"field11_<v>","field12":"field12_<v>","field13":"field13_<v>","field14":"field14_<v>","field15":"field15_<v>","field16":"field16_<v>","field17":"field17_<v>","field18":"field18_<v>","field19":"field19_<v>","field20":null,"field21":"field21_<v>","field22":"field22_<v>","field23":"field23_<v>","field24":"field24_<v>","field25":"field25_<v>","field26":"field26_<v>","field27":"field27_<v>","field28":"field28_<v>","field29":"field29_<v>","field30":null,"field31":"field31_<v>","field32":"field32_<v>","field33":"field33_<v>","field34":"field34_<v>","field35":"field35_<v>","field36":"field36_<v>","field37":"field37_<v>","field38":"field38_<v>","field39":"field39_<v>","field40":null,"field41":"field41_<v>","field42":"field42_<v>","field43":"field43_<v>","field44":"field44_<v>","field45":"field45_<v>","field46":"field46_<v>","field47":"field47_<v>","field48":"field48_<v>","field49":"field49_<v>"}'

import gzip

def write(n):
  with gzip.open(f'json{n}.gz', 'w') as f:
    f.write(b'[')
    for i in range(n):
        if i != 0:
            f.write(b',')
        f.write(s.replace(b'<v>', str(i).encode('ascii')))
    f.write(b']')

write(100000)
```

Then it processes the file in Spark shell with the following command:

```
./bin/spark-shell --conf spark.driver.memory=1g --conf spark.executor.memory=1g  --master "local[1]"

> val schema = "field00 string, field01 string, field02 string, field03 string, field04 string, field05 string, field06 string, field07 string, field08 string, field09 string, field10 string, field11 string, field12 string, field13 string, field14 string, field15 string, field16 string, field17 string, field18 string, field19 string, field20 string, field21 string, field22 string, field23 string, field24 string, field25 string, field26 string, field27 string, field28 string, field29 string, field30 string, field31 string, field32 string, field33 string, field34 string, field35 string, field36 string, field37 string, field38 string, field39 string, field40 string, field41 string, field42 string, field43 string, field44 string, field45 string, field46 string, field47 string, field48 string, field49 string"
> spark.conf.set("spark.sql.json.useUnsafeRow", "false")
> spark.read.schema(schema).option("multiline", "true").json("json100000.gz").selectExpr("sum(hash(struct(*)))").collect()
```

When the flag is off (the current behavior), the query can process 2.5e5 rows but fails to process 3e5 rows. When the flag is on, the query can process 8e5 rows but fails to process 9e5 rows. We can say this change reduces the memory consumption to about 1/3.

### Why are the changes needed?

It reduces the memory requirement of JSON-related query.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

A new JSON unit test with the config flag on.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by chenhao-db** [comment]
@sadikovi please take a look, thanks!

**Comment by LuciferYang** [comment]
Can we add corresponding test scenarios in `DataSourceReadBenchmark` or any other more suitable benchmarks?

**Comment by chenhao-db** [comment]
@LuciferYang I updated the benchmark and PR description. Please take a look.

**Review by sadikovi — APPROVED** [review]

File: sql/core/benchmarks/DataSourceReadBenchmark-results.txt [hunk 1]
```diff
@@ -1,431 +1,438 @@
...
-SQL Parquet Vectorized: DataPageV2                  129            135           6        122.3           8.2      94.9X
-SQL Parquet MR: DataPageV1                         1968           1989          29          8.0         125.1       6.2X
-SQL Parquet MR: DataPageV2                         1913           1916           3          8.2         121.6       6.4X
-SQL ORC Vectorized                                  130            135           6        120.8           8.3      93.7X
-SQL ORC MR                                         1593           1600          10          9.9         101.3       7.7X
-
-OpenJDK 64-Bit Server VM 17.0.11+9-LTS on Linux 6.5.0-1022-azure
+SQL CSV                                           10958          10970          18          1.4         696.7       1.0X
+SQL Json                                          10164          10169           7          1.5         646.2       1.1X
+SQL Json with UnsafeRow                           11113          11137          33          1.4         706.5       1.0X
```
Comment (LuciferYang) [inline sql/core/benchmarks/DataSourceReadBenchmark-results.txt:57]: So, using UnsafeRow is slower than not using it? Is this a negative effect brought about by saving memory?

File: sql/core/benchmarks/DataSourceReadBenchmark-results.txt [hunk 2]
```diff
@@ -1,431 +1,438 @@
-DataSourceReadBenchmark-jdk21-results.txt================================================================================================
+================================================================================================
 SQL Single Numeric Column Scan
 ================================================================================================
 
-OpenJDK 64-Bit Server VM 17.0.11+9-LTS on Linux 6.5.0-1022-azure
+OpenJDK 64-Bit Server VM 17.0.11+9-LTS on Linux 6.5.0-1023-azure
```
Comment (LuciferYang) [inline sql/core/benchmarks/DataSourceReadBenchmark-results.txt:5]: also need a benchmark result from testing with Java 21

Comment (chenhao-db) [reply sql/core/benchmarks/DataSourceReadBenchmark-results.txt:5]: Done.

**Comment by sadikovi** [comment]
cc @dongido001 @HyukjinKwon @LuciferYang for review, thank you!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -4367,6 +4367,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val JSON_USE_UNSAFE_ROW =
+    buildConf("spark.sql.json.useUnsafeRow")
+      .internal()
+      .doc("When set to true, use UnsafeRow to represent struct result in the JSON parser.")
+      .version("4.0.0")
+      .booleanConf
+      .createWithDefault(false)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4376]: Wonder if we should enable this by default. If this benefits us in most of the cases, should be fine to enable it by default

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (chenhao-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4376]: I feel it is better to not enable it by default. The benchmark has shown that applying the change can slow down the JSON scan a bit. If there is enough memory, then saving memory doesn't have enough benefit.

**Comment by chenhao-db** [comment]
@HyukjinKwon thanks! could you help merge it?

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by LuciferYang** [comment]
Sorry, forgot to review this one. Late LGTM

**Pull request closed without merging**